File: iteration2/src/AdvisorCourseRegistrationInterface.java
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

public class AdvisorCourseRegistrationInterface {
    Colors Colors = new Colors();
    private Session session;
    private Scanner scanner;

    private ArrayList<String> selectionCourses;
    private HashMap<String, String> idToSelectionCourses;

    private ArrayList<String> courses;
    private HashMap<String, ArrayList<String>> idToCourses;

    private HashMap<String, String> courseToLectures;
    private HashMap<String, HashMap<String, String>> idToCourseLectures;

    private HashMap<String, String> courseToLabs;
    private HashMap<String, HashMap<String, String>> idToCourseLabs;

    private MessagesInterface messagesInterface;
    private AdvisorInterface advisorInt;
    Object requestsObj;
    JSONArray requestJson;
    Object coursObject;
    JSONArray courseJson;

    public AdvisorCourseRegistrationInterface(Session session, AdvisorInterface advisorInt) {
        this.session = session;
        this.scanner = new Scanner(System.in);

        this.selectionCourses = new ArrayList<String>();
        this.idToSelectionCourses = new HashMap<String, String>();

        this.courses = new ArrayList<String>();
        this.idToCourses = new HashMap<String, ArrayList<String>>();

        this.courseToLectures = new HashMap<String, String>();
        this.idToCourseLectures = new HashMap<String, HashMap<String, String>>();

        this.courseToLabs = new HashMap<String, String>();
        this.idToCourseLabs = new HashMap<String, HashMap<String, String>>();

        try {
            requestsObj = new JSONParser().parse(new FileReader("./jsons/RegistrationRequests.json"));
            requestJson = (JSONArray) requestsObj;
            coursObject = new JSONParser().parse(new FileReader("./jsons/courses.json"));
            courseJson = (JSONArray) coursObject;

        } catch (Exception e) {
            System.out.println(e);
        }

    }

    public void advRegMenu() {

        while (true) {
            System.out.println(Colors.getBOLD() + Colors.getRED() + "\nAdvisor Course Registration Menu\n"
                    + Colors.getRESET() + Colors.getRESET());
            System.out.println(Colors.getYELLOW() + "1" + Colors.getRESET() + ".   Show Students");
            System.out.println(Colors.getYELLOW() + "2" + Colors.getRESET() + ".   Approve/Deny Courses");
            System.out.println(Colors.getYELLOW() + "3" + Colors.getRESET() + ".   Finalize Registration");
            System.out.println(Colors.getYELLOW() + "4" + Colors.getRESET() + ".   Messages Menu");
            System.out.println(Colors.getYELLOW() + "0" + Colors.getRESET() + ".   Go Back to Advisor Menu");
            System.out.print("\n" + Colors.getBLUE() + "--> " + Colors.getRESET() + "What do you want to do?   ");
            System.out.print(Colors.getBLUE());
            char choice = scanner.next().charAt(0);
            System.out.print(Colors.getRESET());

            switch (choice) {
                case '1':
                    showStudents();
                    showStudentsQuestionPart();
                    break;
                case '2':
                    approveCoursesMenu();
                    // return;
                    break;
                case '3':
                    finalizeRegistrationMenu();
                    break;
                case '4':
                    messagesInterface = new MessagesInterface(session);
                    messagesInterface.messagesMenu();
                    break;
                case '0':
                    return; // bu kısımda sıkıntılar var
                default:
                    System.out.println(Colors.getYELLOW() + "Invalid input! Please try again." + Colors.getRESET());
                    advRegMenu();
                    break;
            }
        }

    }

    private void showStudents() {
        Advisor advisor = (Advisor) session.getUser();

        System.out.println(Colors.getBOLD() +
                Colors.getRED() + "\nStudents\n" + Colors.getRESET() + Colors.getRESET());
        int numberOfStudents = 1;
        System.out.println("------------------------------------------------------------");

        System.out.printf("| %6s | %-11s | %-33s |%n", "Number", "Student ID", "Student Name");
        for (Student student : advisor.getStudents()) {
            System.out.println("------------------------------------------------------------");
            System.out.printf("| %-6s | %-11s | %-33s |%n", numberOfStudents, student.getID(),
                    student.getName() + " " + student.getSurname());
            numberOfStudents++;

        }
        System.out.println("------------------------------------------------------------\n");

    }

    private void showStudentsQuestionPart() {
        System.out
                .println(Colors.getYELLOW() + "0" + Colors.getRESET()
                        + ".  Go back to the Advisor Course Registration Menu.\n");
        System.out.print(Colors.getBLUE() + "--> " + Colors.getRESET() + "What do you want to do?   ");

        System.out.print(Colors.getBLUE());
        char choice1 = scanner.next().charAt(0);
        System.out.print(Colors.getRESET());

        if (choice1 == '0') {
            return;
        }
    }

    private void approveCoursesMenu() {
        while (true) {
            Advisor advisor = (Advisor) session.getUser();

            showStudents();

            System.out.println(
                    Colors.getYELLOW() + "0" + Colors.getRESET()
                            + ".  Go back to the Advisor Course Registration Menu.\n");
            System.out.print(Colors.getBLUE() + "--> " + Colors.getRESET() + "Select student for approval: ");
            System.out.print(Colors.getBLUE());
            int choice = scanner.nextInt();
            System.out.print(Colors.getRESET());
            if (choice == 0) {
                return;
            }

            Student student = advisor.getStudents().get(choice - 1);
            System.out.println();
            System.out.println(Colors.getGREEN() + "Student ID: " + student.getID() + "\n" + "Student Name: "
                    + student.getName() + " " + student.getSurname() + Colors.getRESET() + "\n");
            String studentID = student.getID();
            showStudentsRequested(studentID);
        }

    }

    private void showStudentsRequested(String studentID) {
        int numberOfCourses = 1;

        for (Object requestObj : requestJson) {
            JSONObject request = (JSONObject) requestObj;
            String requestID = (String) request.get("StudentID");
            if (studentID.equals(requestID)) {
                JSONArray selectedCourses = (JSONArray) request.get("SelectedCourses");
                JSONArray selectedLectures = (JSONArray) request.get("SelectedLectures");
                JSONArray selectedLabs = (JSONArray) request.get("SelectedLabs");

                for (Object courseObj : selectedCourses) {
                    String course = (String) courseObj;
                    for (Object lectureObj : selectedLectures) {
                        String lecture = (String) lectureObj;
                        if (lecture.contains(course)) {
                            courseToLectures.put(course, lecture);
                        }
                    }

                    for (Object labObj : selectedLabs) {
                        String lab = (String) labObj;
                        if (lab.contains(course)) {
                            courseToLabs.put(course, lab);
                        }
                    }
                    courses.add(course);

                    numberOfCourses++;
                }

                // mapping student id to courses
                idToCourses.put(studentID, courses);
                idToCourseLectures.put(studentID, courseToLectures);
                idToCourseLabs.put(studentID, courseToLabs);

                // resetting for new student
                courses = new ArrayList<String>();
                courseToLectures = new HashMap<String, String>();
                courseToLabs = new HashMap<String, String>();

                // print lectures and labs
                for (int i = 0; i < numberOfCourses - 1; i++) {
                    System.out
                            .println(Colors.getYELLOW() + (i + 1) + Colors.getRESET() + ". "
                                    + idToCourses.get(studentID).get(i));
                    if (idToCourseLectures.get(studentID).get(idToCourses.get(studentID).get(i)) != null) {
                        System.out.println(Colors.getYELLOW() + "Lecture: " + Colors.getRESET()
                                + idToCourseLectures.get(studentID).get(idToCourses.get(studentID).get(i)));

                    }
                    if (idToCourseLabs.get(studentID).get(idToCourses.get(studentID).get(i)) != null) {
                        System.out.println(Colors.getYELLOW() + "Lab: " + Colors.getRESET()
                                + idToCourseLabs.get(studentID).get(idToCourses.get(studentID).get(i)));

                    }
                    System.out.println("----------------------------------");
                }

            }
        }
        approveCourse(studentID);

    }

    private void approveCourse(String studentID) {
        if (idToCourses.isEmpty()) {
            System.out
                    .println(Colors.getYELLOW() + "No courses to approve! Please select different student!"
                            + Colors.getRESET());
            return;
        }

        System.out.println(
                Colors.getYELLOW() + "0" + Colors.getRESET() + ".  Go back to the Advisor Course Registration Menu.");
        System.out.print(Colors.getBLUE() + "\n--> " + Colors.getRESET() + "Select course to approve: ");

        while (true) {
            System.out.print(Colors.getBLUE());
            int choice = scanner.nextInt();
            System.out.print(Colors.getRESET());
            if (choice == 0) {
                break;
            }

            String course = idToCourses.get(studentID).get(choice - 1);
            selectionCourses.add(course);
            System.out.println(Colors.getGREEN() + course + " is approved!" + Colors.getRESET());
            System.out.println(
                    Colors.getBLUE() + " If your selection is finished, you can press 0. " + Colors.getRESET());
            System.out.print(Colors.getRED() + " OR " + Colors.getBLUE() + "\n--> " + Colors.getRESET()
                    + "Select course to approve: ");
        }
        saveApprovel(studentID);
        // send notification to student
        String receiverID = studentID;
        String description = "Advisor has approved your courses!";
        String senderID = session.getUser().getID();
        Notification notification = new Notification(receiverID, description, senderID);
        notification.sendNotification(senderID);

    }

    private void saveApprovel(String studentID) {
        for (Object requestObj : requestJson) {
            JSONObject request = (JSONObject) requestObj;
            String requestID = (String) request.get("StudentID");
            if (studentID.equals(requestID)) {
                JSONArray approvedCoursesJsonArray = (JSONArray) request.get("ApprovedCourses");
                JSONArray selectedCourses = (JSONArray) request.get("SelectedCourses");

                JSONArray approvedLectureJsonArray = (JSONArray) request.get("ApprovedLectures");
                JSONArray selectedLectures = (JSONArray) request.get("SelectedLectures");

                JSONArray approvedLabJsonArray = (JSONArray) request.get("ApprovedLabs");
                JSONArray selectedLabs = (JSONArray) request.get("SelectedLabs");

                HashMap<String, String> courseToLabsTemp = idToCourseLabs.get(studentID);
                HashMap<String, String> courseToLectureTemp = idToCourseLectures.get(studentID);

                for (String course : selectionCourses) {
                    approvedCoursesJsonArray.add(course);
                    if (courseToLabsTemp.get(course) != null) {
                        approvedLabJsonArray.add(courseToLabsTemp.get(course));
                    }
                    if (courseToLectureTemp.get(course) != null) {
                        approvedLectureJsonArray.add(courseToLectureTemp.get(course));
                    }
                }
                selectedCourses.clear();
                selectedLectures.clear();
                selectedLabs.clear();

            }
        }

        try {
            PrintWriter pw = new PrintWriter("./jsons/RegistrationRequests.json");
            pw.write(requestJson.toJSONString());

            pw.flush();
            pw.close();

        } catch (Exception e) {
        }

    }

    private void finalizeRegistrationMenu() {
        showStudents();
        System.out.println(
                Colors.getYELLOW() + "0" + Colors.getRESET() + ".  Go back to the Advisor Course Registration Menu.\n");
        System.out.print(Colors.getBLUE() + "--> " + Colors.getRESET() + "Select student for finalization: ");
        System.out.print(Colors.getBLUE());
        int choice = scanner.nextInt();
        System.out.print(Colors.getRESET());
        Advisor advisor = (Advisor) session.getUser();

        Student curStudent = advisor.getStudents().get(choice - 1);
        Object forDelete = new Object();
        for (Object requestObj : requestJson) {
            forDelete = requestObj;
            ArrayList<JSONObject> finalizedArrayList = new ArrayList<JSONObject>();

            JSONObject request = (JSONObject) requestObj;
            String requestID = (String) request.get("StudentID");
            if (!curStudent.getID().equals(requestID)) {
                continue;
            }
            JSONArray approvedCoursesJsonArray = (JSONArray) request.get("ApprovedCourses");
            if (approvedCoursesJsonArray.isEmpty()) {
                System.out.println(Colors.getYELLOW() + "No courses to finalize! Please select different student!"
                        + Colors.getRESET());
                return;
            }
            for (Object courseObj : approvedCoursesJsonArray) {
                String course = (String) courseObj;
                for (Object courseListObject : courseJson) {
                    JSONObject courseList = (JSONObject) courseListObject;
                    String courseID = (String) courseList.get("CourseID");
                    if (course.equals(courseID)) {
                        String newCourseID = (String) courseList.get("CourseID");
                        String newCourseName = (String) courseList.get("CourseName");
                        String newCourseType = (String) courseList.get("Type");
                        Long newCourseCredit = (Long) (courseList.get("Credit"));
                        Long newCourseSemester = (Long) courseList.get("Semester");
                        Double newCourseGrade = 0.0;

                        JSONObject newCourse = new JSONObject();
                        newCourse.put("CourseID", newCourseID);
                        newCourse.put("CourseName", newCourseName);
                        newCourse.put("Type", newCourseType);
                        newCourse.put("Credit", newCourseCredit);
                        newCourse.put("Semester", newCourseSemester);
                        newCourse.put("Grade", newCourseGrade);

                        finalizedArrayList.add(newCourse);

                    }
                }

            }

            String filename = requestID + ".json";
            try {
                Object studentObj = new JSONParser().parse(new FileReader("./jsons/student/" + filename));
                JSONObject student = (JSONObject) studentObj;
                JSONObject transcriptObj = (JSONObject) student.get("Transcript");
                JSONArray semArray = (JSONArray) transcriptObj.get("Semester");

                JSONArray takenCourses = (JSONArray) student.get("TakenCourses");
                String studentID = (String) student.get("ID");
                HashMap<String, String> courseToLabsTemp = idToCourseLabs.get(studentID);
                HashMap<String, String> courseToLectureTemp = idToCourseLectures.get(studentID);

                for (Object courseObj : approvedCoursesJsonArray) {
                    String course = (String) courseObj;
                    takenCourses.add(course);
                    if (courseToLabsTemp == null) {
                        continue;
                    }
                    if (courseToLabsTemp.get(course) != null) {
                        takenCourses.add(courseToLabsTemp.get(course));
                    }
                    if (courseToLectureTemp.get(course) != null) {
                        takenCourses.add(courseToLectureTemp.get(course));
                    }
                }

                JSONObject newCourses = new JSONObject();

                newCourses.put("Courses", finalizedArrayList);

                semArray.add(newCourses);

                PrintWriter studentPw = new PrintWriter("./jsons/student/" + filename);
                studentPw.write(student.toJSONString());

                studentPw.flush();
                studentPw.close();

                // send notification to student
                String receiverID = studentID;
                String description = "Advisor has finalized your courses!";
                String senderID = advisor.getID();
                Notification notification = new Notification(receiverID, description, senderID);
                notification.sendNotification(senderID);

                // statement ekle bu isimli öğrencinin dersleri finilization edildi diye advisor
                // course reg menuye dönüyorsunuz tarzı aynı logindeki gibi bir senaryo olabilir
                System.out.println(Colors.getGREEN() + "Courses of student " + studentID + " are finalized!" + Colors.getRESET());

            } catch (Exception e) {
                System.out.println(e);
            }

        }
        try {
            requestJson.remove(forDelete);
            PrintWriter requestPw = new PrintWriter("./jsons/RegistrationRequests.json");
            requestPw.write(requestJson.toJSONString());
            requestPw.flush();
            requestPw.close();
        } catch (Exception e) {
        }
    }

}

File: iteration3/src/advisor_course_registration_interface.py
import json
from notification import Notification
from colors import Colors
from session import Session
import logging
from datetime import datetime


# color need fix
class AdvisorCourseRegistrationInterface:
    logging.basicConfig(
        filename="./logs/app.log",
        filemode="w",
        format="%(name)s - %(levelname)s - %(message)s",
        level=logging.INFO,
    )

    def __init__(self, session: Session):
        self
        self.__colors = Colors()
        self.__session = session

        self.__selection_courses: list = []
        self.__id_to_courses = {}

        self.__id_to_course_lectures = {}

        self.__id_to_course_labs = {}

        # self.__messages_interface = MessagesInterface(self.__session)

        try:
            self.__request_file = json.load(open("./jsons/RegistrationRequests.json"))
            self.__courses_file = json.load(open("./jsons/courses.json"))

        except Exception as e:
            logging.error("time: {}, User {} error {}".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), self.__session.get_user().get_id(), e))
            print(e)

    def adv_reg_menu(self):
        while True:
            print(
                f"{self.__colors.get_bold()}{self.__colors.get_red()}\n>> Advisor Course Registration Menu{self.__colors.get_reset()}{self.__colors.get_reset()}"
            )
            print(
                f"{self.__colors.get_yellow()}1{self.__colors.get_reset()}.   Show Students"
            )
            print(
                f"{self.__colors.get_yellow()}2{self.__colors.get_reset()}.   Approve/Deny Courses"
            )
            print(
                f"{self.__colors.get_yellow()}3{self.__colors.get_reset()}.   Finalize Registration"
            )
            print(
                f"{self.__colors.get_yellow()}4{self.__colors.get_reset()}.   Messages Menu"
            )
            print(
                f"{self.__colors.get_yellow()}0{self.__colors.get_reset()}.   Go Back to Advisor Menu"
            )
            print(
                "\n"
                + self.__colors.get_blue()
                + "--> "
                + self.__colors.get_reset()
                + "What do you want to do?   ",
                end="",
            )
            print(self.__colors.get_blue(), end="")
            choice = int(input())
            print(self.__colors.get_reset(), end="")
            if choice == 1:
                self.__show_students()
                self.__show_students_question_part()
            elif choice == 2:
                self.__approve_courses_menu()
            elif choice == 3:
                self.__finalize_registration_menu()
            elif choice == 4:
                self.__messages_interface.messages_menu()
            elif choice == 0:
                return
            else:
                print(
                    f"{self.__colors.get_yellow()}Invalid input! Please give a number!{self.__colors.get_reset()}"
                )

    def __show_students(self):
        advisor = self.__session.get_user()

        print(
            f"{self.__colors.get_bold()}{self.__colors.get_red()}\n>>> Students{self.__colors.get_reset()}{self.__colors.get_reset()}"
        )
        number_of_students = 1
        print("------------------------------------------------------------")

        print("| %6s | %-11s | %-33s |" % ("Number", "Student ID", "Student Name"))
        for student in advisor.get_students():
            print("------------------------------------------------------------")
            print(
                "| %-6s | %-11s | %-33s |"
                % (
                    number_of_students,
                    student.get_id(),
                    student.get_name() + " " + student.get_surname(),
                )
            )
            number_of_students += 1

        print("------------------------------------------------------------\n")

    def __show_students_question_part(self):
        print(
            f"{self.__colors.get_yellow()}0{self.__colors.get_reset()}.  Go back to the Advisor Course Registration Menu.\n"
        )
        print(
            f"{self.__colors.get_blue()}--> {self.__colors.get_reset()}What do you want to do?   ",
            end="",
        )

        choice1 = input()

        if choice1 == "0":
            return

    def __approve_courses_menu(self):
        while True:
            advisor = self.__session.get_user()

            self.__show_students()

            print(
                f"{self.__colors.get_yellow()}0{self.__colors.get_reset()}.  Go back to the Advisor Course Registration Menu.\n"
            )
            print(
                f"{self.__colors.get_blue()}--> {self.__colors.get_reset()}Select student for approval: ",
                end="",
            )

            choice = int(input())
            if choice == 0:
                return

            student = advisor.get_students()[choice - 1]
            print()
            print(
                f"{self.__colors.get_green()}Student ID: {student.get_id()}\nStudent Name: {student.get_name()} {student.get_surname()}{self.__colors.get_reset()}\n"
            )
            student_id = student.get_id()
            self.__show_students_requested(student_id)

    def __show_students_requested(self, student_id):
        print(
            f"{self.__colors.get_bold()}{self.__colors.get_red()}\n>>> Requested Courses{self.__colors.get_reset()}{self.__colors.get_reset()}"
        )
        if not self.__request_file:
            print(
                f"{self.__colors.get_yellow()}There is no requested courses!{self.__colors.get_reset()}"
            )
            return
        for request_obj in self.__request_file:
            request_id = request_obj["StudentID"]
            if student_id != request_id:
                continue

            course = request_obj["SelectedCourses"]
            lecture = request_obj["SelectedLectures"]
            lab = request_obj["SelectedLabs"]

            # mapping for id to
            self.__id_to_courses = {student_id: course}
            self.__id_to_course_lectures = {student_id: lecture}
            self.__id_to_course_labs = {student_id: lab}

            for course in self.__id_to_courses[student_id]:
                print(
                    f"{self.__colors.get_yellow()}{course}{self.__colors.get_reset()}"
                )
                if self.__id_to_course_lectures[student_id] is not None:
                    for lecture in self.__id_to_course_lectures[student_id]:
                        if lecture.count(course) > 0:
                            print(
                                f"{self.__colors.get_yellow()}{lecture}{self.__colors.get_reset()}"
                            )
                if self.__id_to_course_labs[student_id] is not None:
                    for lab in self.__id_to_course_labs[student_id]:
                        if lab.count(course) > 0:
                            print(
                                f"{self.__colors.get_yellow()}{lab}{self.__colors.get_reset()}"
                            )
                self.__approve_course(student_id)

    def __approve_course(self, student_id):
        if not self.__id_to_courses:
            print(
                f"{self.__colors.get_yellow()}No courses to approve! Please select a different student!{self.__colors.get_reset()}"
            )
            return

        print(
            f"{self.__colors.get_yellow()}0{self.__colors.get_reset()}.  Go back to the Advisor Course Registration Menu."
        )
        print(
            f"{self.__colors.get_blue()}--> {self.__colors.get_reset()}Select course to approve: ",
            end="",
        )

        course_len = len(self.__id_to_courses[student_id])
        while True:
            choice = int(input())
            if choice == 0:
                break
            if choice > course_len:
                print(
                    f"{self.__colors.get_red()} invalid input {self.__colors.get_reset()}"
                )
                return
            course = self.__id_to_courses[student_id][choice - 1]
            if (
                self.__selection_courses is not None
                and course in self.__selection_courses
            ):
                print(
                    f"{self.__colors.get_yellow()}You have already approved this course! Press 0{self.__colors.get_reset()}"
                )
                continue
            else:
                self.__selection_courses.append(course)

            print(
                f"{self.__colors.get_green()}{course} is approved!{self.__colors.get_reset()}"
            )
            print(
                f"{self.__colors.get_blue()}If your selection is finished, you can press 0.{self.__colors.get_reset()}"
            )
            print(
                f"{self.__colors.get_red()}OR{self.__colors.get_blue()}\n--> Select course to approve: ",
                end="",
            )

        self.__save_approval(student_id)

        # send notification to student
        receiver_id = student_id
        description = "Advisor has approved your courses!"
        sender_id = self.__session.get_user().get_id()
        notification = Notification(receiver_id, description, sender_id)
        notification.send_notification(sender_id)

    def __save_approval(self, student_id):
        for request in self.__request_file:
            if request["StudentID"] != student_id:
                continue

            if self.__selection_courses is None:
                print(f"{self.__colors.get_red()} There is no approved courses ")
                break

            request["ApprovedCourses"] = self.__selection_courses

            for course in self.__selection_courses:
                if self.__id_to_course_lectures[student_id] is not None:
                    for lecture in self.__id_to_course_lectures[student_id]:
                        if lecture.count(course) > 0:
                            request["ApprovedLectures"].append(lecture)

                    for lab in self.__id_to_course_labs[student_id]:
                        if lab.count(course) > 0:
                            request["ApprovedLabs"].append(lab)

            request["SelectedCourses"] = []
            request["SelectedLectures"] = []
            request["SelectedLabs"] = []

        logging.info(
            "time: {}, User {} approve courses {}".format(
                datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
                self.__session.get_user().get_id(),
                student_id,
            )
        )

        try:
            with open("./jsons/RegistrationRequests.json", "w") as request_file:
                json.dump(self.__request_file, request_file)

        except Exception as e:
            logging.error("time: {}, User {} error {}".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), self.__session.get_user().get_id(), e))
            print(e)

    def __finalize_registration_menu(self):
        while True:
            self.__show_students()
            print(
                f"{self.__colors.get_yellow()} Select student to finalize registration {self.__colors.get_reset()}"
            )

            choice = int(input())
            if choice == 0:
                return

            student = self.__session.get_user().get_students()[choice - 1]

            try:
                with open(
                    "./jsons/student/" + student.get_id() + ".json", "r"
                ) as student_file:
                    student_data = json.load(student_file)
            except Exception as e:
                logging.error("time: {}, User {} error {}".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), self.__session.get_user().get_id(), e))
                print(e)

            for request in self.__request_file:
                if request["StudentID"] != student.get_id():
                    continue

                if request["ApprovedCourses"] is None:
                    print(f"{self.__colors.get_red()} There is no approved courses ")
                    break

                student_data["Transcript"]["Semester"].append({"Courses": []})

                for curr_course in request["ApprovedCourses"]:
                    for course in self.__courses_file:
                        if course["CourseID"] == curr_course:
                            newCourse = {
                                "CourseID": course["CourseID"],
                                "CourseName": course["CourseName"],
                                "Credit": course["Credit"],
                                "Type": course["Type"],
                                "Semester": course["Semester"],
                                "Grade": str(0),
                            }
                            student_data["Transcript"]["Semester"][-1][
                                "Courses"
                            ].append(newCourse)

                request.clear()
                logging.info(
                    "time: {}, User {} finalized courses for {}.".format(
                        datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
                        self.__session.get_user().get_id(),
                        student.get_id(),
                    )
                )

            try:
                with open(
                    "./jsons/student/" + student.get_id() + ".json", "w"
                ) as student_file:
                    json.dump(student_data, student_file)

                with open("./jsons/RegistrationRequests.json", "w") as request_file:
                    json.dump(self.__request_file, request_file)
            except Exception as e:
                logging.error("time: {}, User {} error {}".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), self.__session.get_user().get_id(), e))
                print(e)
